dialer.go

```diff
@@ -339,8 +399,29 @@ func (d *Dialer) Dial(ctx context.Context, icn string, opts ...DialOption) (conn
 		return nil, ErrDialerClosed
 	default:
 	}
+	cfg := d.defaultDialConfig
+	for _, opt := range opts {
+		opt(&cfg)
+	}
+
+	// Resolve the instance connection name to a ConnName struct.
+	// Note: icn may be a domain name that resolves to an instance connection name.
+	cn, err := d.resolver.Resolve(ctx, icn)
```

Comment:
We should count DNS lookup successes and errors. Does this make more sense as a label on our connect count metric, or is DNS resolution + status its own separate counter.